tutorials/v7/climate/climate_tutorial.C

```diff
@@ -0,0 +1,249 @@
+      if (nFields != 9) {
+         nSkipped++;
+         continue;
+      }
+      *fieldCountry = country;
+      *fieldCity = city;
+
+      ntuple->Fill();
+
+      if (++nRecords % 1000000 == 0)
+         std::cout << "  ... converted " << nRecords << " records" << std::endl;
+   }
+
+   // Display the total time to process the data
+   std::cout << nSkipped << " records skipped" << std::endl;
+   std::cout << nRecords << " records processed" << std::endl;
+
+   auto t2 = Clock::now();
+   std::cout << std::endl
+             << "Processing Time: "
+             << std::chrono::duration_cast<std::chrono::seconds>(t2 - t1).count()
+             << " seconds\n" << std::endl;
+}
+
+
+void Analyze() {
+   // Get a unique pointer to an empty RNTuple model
+   auto model = RNTupleModel::Create();
+   // Define only the necessary fields for reading
+   model->MakeField<float>("AverageTemperature");
+   model->MakeField<float>("AverageTemperatureUncertainty");
+   model->MakeField<std::uint32_t>("Year");
+   model->MakeField<std::uint32_t>("Month");
+   model->MakeField<std::uint32_t>("Day");
+   model->MakeField<std::string>("City");
+
+   // Create an ntuple and attach the read model to it
+   auto ntuple = RNTupleReader::Open(std::move(model), "GlobalTempData", kNTupleFileName);
+   // Overview of the ntuple and list of fields.
+   ntuple->PrintInfo();
+   std::cout << std::endl
+             << "The tenth entry in JSON format:" << std::endl;
+   ntuple->Show(9);
+
+   // Create a RDataframe by wrapping around NTuple
+   auto df = ROOT::Experimental::MakeNTupleDataFrame("GlobalTempData", kNTupleFileName);
+   // Display the minimum and maximum temperature from the dataset
+   auto minimum_value = df.Min("AverageTemperature");
+   cout << "\nThe Minimum temperature is: " << *minimum_value << std::endl;
+   auto max_value = df.Max("AverageTemperature");
+   cout << "The Maximum temperature is: " << *max_value << std::endl;
+
+   // Functions to filter by each season from date formatted "1944-12-01"
+   auto fnWinter = [](int month) { return month == 12 || month == 1  || month == 2;  };
+   auto fnSpring = [](int month) { return month == 3  || month == 4  || month == 5;  };
+   auto fnSummer = [](int month) { return month == 6  || month == 7  || month == 8;  };
+   auto fnFall   = [](int month) { return month == 9  || month == 10 || month == 11; };
+
+   // Create a RDataFrame per season by filtering ntupleToDF and display the count for each season
+   auto dfWinter = df.Filter(fnWinter, {"Month"});
+   cout << "\nThe count for Winter: " << *dfWinter.Count() << std::endl;
+   auto dfSpring = df.Filter(fnSpring, {"Month"});
+   cout << "The count for Spring: " << *dfSpring.Count() << std::endl;
+   auto dfSummer = df.Filter(fnSummer, {"Month"});
+   cout << "The count for Summer: " << *dfSummer.Count() << std::endl;
+   auto dfFall = df.Filter(fnFall, {"Month"});
+   cout << "The count for Fall: " << *dfFall.Count() << std::endl;
```

Comment:
although of less relevance for this dataset size, one of the major features of RDF is laziness:

```cpp
*df.Count();
*df.Count();
```
performs, necessarily, two loops over the data, but

```cpp
auto c1 = df.Count();
auto c2 = df.Count();
*c1, *c2;
```
performs only one loop.

We should definitely teach/show the latter approach, which makes a _huge_ runtime difference for real-world usecases. Please refactor so that the results of all `Count`, `Histo1D`, `Min` and `Max` are all booked before one of them is used. Maybe even point out `"no computation ran so far. a single event loop runs here"` before the first access to one of the results.

Reply:
@eguiraud I'm not sure I follow.  In this case, we count on different filters (`dfSummer`, `dfFall`, etc), and the call to `*df<Season>.Count()` is only issued once.

Do you mean as a general pattern we should store the result in a variable because otherwise copy & paste to other macros could eventually lead to `Count()` being called multiple times?

Reply:
> Do you mean as a general pattern we should store the result in a variable because otherwise copy & paste to other macros could eventually lead to Count() being called multiple times?

No no, I mean that this loops over data twice:

```c++
auto dfSummer = df.Filter(fnSummer, {"Month"});
cout << "The count for Summer: " << *dfSummer.Count() << std::endl;
auto dfFall = df.Filter(fnFall, {"Month"});
cout << "The count for Fall: " << *dfFall.Count() << std::endl;
```

while this would loop over data once and we should encourage/show/teach this pattern instead:

```c++
auto dfSummer = df.Filter(fnSummer, {"Month"});
auto c1 = dfSummer.Count();
auto dfFall = df.Filter(fnFall, {"Month"});
auto c2 = dfFall.Count();
cout << "The count for Summer: " << *c1 << std::endl;
cout << "The count for Fall: " << *c2 << std::endl;
```

The main idea in RDF is that you want to first book all actions and then access the results (and the first access to any of the booked results will trigger the one single event loop that produces them all).

Reply:
Oh, I see. Thanks for the clarification! @JohnYoon13 can you rewrite accordingly?